x/staking/keeper/delegation.go

```diff
@@ -1037,6 +1196,70 @@ func (k Keeper) CompleteRedelegation(
 	return balances, nil
 }
 
+func redelegationEntryArrayIndex(red types.Redelegation, id uint64) (index int, found bool) {
+	for i, entry := range red.Entries {
+		// we find the entry with the right ID
+		if entry.Id == id {
+			return i, true
+		}
+	}
+
+	return 0, false
+}
+
+func (k Keeper) PutRedelegationEntryOnHold(ctx sdk.Context, id uint64) (found bool) {
+	red, found := k.GetRedelegationByUnbondingOpId(ctx, id)
+	if !found {
+		return false
+	}
+
+	i, found := redelegationEntryArrayIndex(red, id)
+	if !found {
+		return false
+	}
+
+	red.Entries[i].OnHold = true
+
+	k.SetRedelegation(ctx, red)
+
+	return true
+}
+
+// This can be called to complete the unbonding of an unbonding delegation entry that was previously
+// stopped by the BeforeUnbondingDelegationEntryComplete hook in CompleteUnbonding
+func (k Keeper) RedelegationCanComplete(ctx sdk.Context, id uint64) (found bool, err error) {
+	red, found := k.GetRedelegationByUnbondingOpId(ctx, id)
+	if !found {
+		return false, nil
+	}
+
+	i, found := redelegationEntryArrayIndex(red, id)
+	if !found {
+		return false, nil
+	}
+
+	if !red.Entries[i].IsMature(ctx.BlockHeader().Time) {
+		// If not matured, set onHold to false
+		red.Entries[i].OnHold = false
+	} else {
+		// If matured, complete it.
+		// Remove entry
+		red.RemoveEntry(int64(i))
+		// Remove from the UBDByUnbondingOp index
+		k.DeleteREDByUnbondingOpIndex(ctx, red.Entries[i].Id)
```

Comment:
The delete should be done also on the other branch. In general, it can be done in `UnbondingOpCanComplete`